urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/Base64.java

```diff
@@ -0,0 +1,1977 @@
+      encoded.put(enc4);
+    }   // end input remaining
+  }
+
+  /**
+   * Performs Base64 encoding on the
+   * <code>raw</code> ByteBuffer, writing it to the
+   * <code>encoded</code> CharBuffer. This is an experimental feature. Currently
+   * it does not pass along any options (such as {@link #DO_BREAK_LINES} or {@link #GZIP}.
+   *
+   * @param raw input buffer
+   * @param encoded output buffer
+   * @since 2.3
+   */
+  public static void encode(java.nio.ByteBuffer raw, java.nio.CharBuffer encoded) {
+    byte[] raw3 = new byte[3];
+    byte[] enc4 = new byte[4];
+
+    while (raw.hasRemaining()) {
+      int rem = Math.min(3, raw.remaining());
+      raw.get(raw3, 0, rem);
+      Base64.encode3to4(enc4, raw3, rem, Base64.NO_OPTIONS);
+      for (int i = 0 ; i < 4 ; i++) {
+        encoded.put((char) (enc4[i] & 0xFF));
+      }
+    }   // end input remaining
+  }
+
+  /**
+   * Serializes an object and returns the Base64-encoded version of that
+   * serialized object.
+   *
+   * <p>As of v 2.3, if the object cannot be serialized or there is another
+   * error, the method will throw an java.io.IOException. <b>This is new to
+   * v2.3!</b> In earlier versions, it just returned a null value, but in
+   * retrospect that's a pretty poor way to handle it.</p>
+   *
+   * The object is not GZip-compressed before being encoded.
+   *
+   * @param serializableObject The object to encode
+   * @return The Base64-encoded object
+   * @throws java.io.IOException if there is an error
+   * @throws NullPointerException if serializedObject is null
+   * @since 1.4
+   */
+  public static String encodeObject(java.io.Serializable serializableObject)
+          throws java.io.IOException {
+    return encodeObject(serializableObject, NO_OPTIONS);
+  }   // end encodeObject
+
+  /**
+   * Serializes an object and returns the Base64-encoded version of that
+   * serialized object.
+   *
+   * <p>As of v 2.3, if the object cannot be serialized or there is another
+   * error, the method will throw an java.io.IOException. <b>This is new to
+   * v2.3!</b> In earlier versions, it just returned a null value, but in
+   * retrospect that's a pretty poor way to handle it.</p>
+   *
+   * The object is not GZip-compressed before being encoded. <p> Example options:
+   * <pre>
+   *   GZIP: gzip-compresses object before encoding it.
+   *   DO_BREAK_LINES: break lines at 76 characters
+   * </pre> <p> Example:
+   * <code>encodeObject( myObj, Base64.GZIP )</code> or <p> Example:
+   * <code>encodeObject( myObj, Base64.GZIP | Base64.DO_BREAK_LINES )</code>
+   *
+   * @param serializableObject The object to encode
+   * @param options Specified options
+   * @return The Base64-encoded object
+   * @see Base64#GZIP
+   * @see Base64#DO_BREAK_LINES
+   * @throws java.io.IOException if there is an error
+   * @since 2.0
+   */
+  public static String encodeObject(java.io.Serializable serializableObject, int options)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.